vllm/worker/hpu_enc_dec_model_runner.py

```diff
@@ -0,0 +1,606 @@
+
+            attn_metadata.cross_block_list = block_list
+            attn_metadata.cross_block_groups = block_groups
+            attn_metadata.cross_block_usage = block_usage
+
+        encoder_seq_lens_tensor = self._list_to_int32_tensor(encoder_seq_lens)
+        attn_metadata.encoder_seq_lens = encoder_seq_lens
+        attn_metadata.encoder_seq_lens_tensor = encoder_seq_lens_tensor
+
+        return attn_metadata
+
+    def profile_run(self) -> None:
+        num_layers = self.model_config.get_num_layers(self.parallel_config)
+        kv_caches = [None] * num_layers
+        max_batch_size = self.max_num_prefill_seqs
+        _, max_seq_len = self.bucketing_ctx.get_max_prompt_shape()
+        max_seq_len = min(self.max_num_batched_tokens // max_batch_size,
+                          max_seq_len)
+
+        self.warmup_scenario(max_batch_size, max_seq_len, True, kv_caches,
+                             False)
+        return
+
+    def warmup_scenario(self,
+                        batch_size,
+                        seq_len,
+                        is_prompt,
+                        kv_caches,
+                        is_pt_profiler_run=False,
+                        is_lora_profile_run=False,
+                        temperature=0) -> None:
+        use_graphs = self._use_graphs(batch_size, seq_len, is_prompt)
+        scenario_name = ("warmup_"
+                         f"{'prompt' if is_prompt else 'decode'}_"
+                         f"bs{batch_size}_"
+                         f"seq{seq_len}_"
+                         f"graphs{'T' if use_graphs else 'F'}")
+        self.profiler.start('internal', scenario_name)
+        times = 3 if use_graphs or is_pt_profiler_run else 1
+        if is_prompt:
+            seqs = [
+                self.create_dummy_seq_group_metadata(i, seq_len, is_prompt)
+                for i in range(batch_size)
+            ]
+        else:
+            # FIXME: seq_len is actually number of blocks
+            blocks = [seq_len // batch_size for _ in range(batch_size)]
+            blocks[0] += seq_len % batch_size
+            seqs = [
+                self.create_dummy_seq_group_metadata(i,
+                                                     b * self.block_size - 1,
+                                                     is_prompt)
+                for i, b in enumerate(blocks)
+            ]
+        torch.hpu.synchronize()
+        profiler = None
+        if is_pt_profiler_run and self.is_driver_worker:
+            profiler = setup_profiler()
+            profiler.start()
+        for _ in range(times):
+            inputs = self.prepare_model_input(seqs)
+            self.execute_model(inputs, kv_caches, warmup_mode=True)
+            torch.hpu.synchronize()
+            if profiler:
+                profiler.step()
+        if profiler:
+            profiler.stop()
+        self.profiler.end()
+        gc.collect()
+
+    def create_dummy_seq_group_metadata(self,
```

Comment:
do we need to define new function? can't we derive function from HPUModelRunnerBase?
It looks like a 1:1 copy

Reply:
They are different. In encoder-decoder scenario, additional cross-attention blocks and encoder seq data should be constructed to the `SequenceGroupMetadata`.